### PR TITLE
Add error for `...` in expressions

### PR DESCRIPTION
**Pull request by Badel2**
Follow-up to https://github.com/rust-lang/rust/pull/44709
Tracking issue: https://github.com/rust-lang/rust/issues/28237

* Using `...` in expressions was a warning, now it's an error
* The error message suggests using `..` or `..=` instead, and explains the difference
* Updated remaining occurrences of `...` to `..=`

r? petrochenkov

### CODE REVIEW

**Comment by petrochenkov** [comment]
I'm sad every time I have to @bors r+ this

File: src/libsyntax/util/parser.rs [hunk 1]
```diff
@@ -106,7 +106,8 @@ impl AssocOp {
             Token::OrOr => Some(LOr),
             Token::DotDot => Some(DotDot),
             Token::DotDotEq => Some(DotDotEq),
-            Token::DotDotDot => Some(DotDotEq), // remove this after SNAP
+            // DotDotDot is no longer supported, but we need some way to display the error
+            Token::DotDotDot => Some(DotDotEq),
```
Comment (Badel2) [inline src/libsyntax/util/parser.rs]: My main concern is that in order to show the error we need to parse the expression, so here I'm parsing `Token::DotDotDot` as `AssocOp::DotDotEq`, and in src/libsyntax/parse/parser.rs I check for `Token::DotDotDot` and `AssocOp::DotDotEq` to show the error. An alternative would be returning `None` here but then the expression would be incomplete (adding the generic "expected one of" error).

Comment (petrochenkov) [reply src/libsyntax/util/parser.rs]: >I'm parsing `Token::DotDotDot` as `AssocOp::DotDotEq`

That's exactly what is needed for good error recovery.

Comment (Badel2) [reply src/libsyntax/util/parser.rs]: Nice, thanks!

**Comment by kennytm** [comment]
Several `run-pass/pretty` tests failed, and one of them caused ICE.

<details><summary>run-pass/range_inclusive.rs</summary>

```
[01:22:25] ---- [pretty] run-pass/range_inclusive.rs stdout ----
[01:22:25] 	
[01:22:25] error: pretty-printing failed in round 1 revision None
[01:22:25] status: exit code: 101
[01:22:25] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "-" "-Zunstable-options" "--unpretty" "normal" "--target" "x86_64-unknown-linux-gnu" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/range_inclusive.stage2-x86_64-unknown-linux-gnu.pretty.aux" "-Crpath" "-O" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
[01:22:25] stdout:
[01:22:25] ------------------------------------------
[01:22:25] 
[01:22:25] ------------------------------------------
[01:22:25] stderr:
[01:22:25] ------------------------------------------
[01:22:25] error: `...` syntax cannot be used in expressions
[01:22:25]   --> <anon>:20:56
[01:22:25]    |
[01:22:25] 20 | fn return_range_to() -> RangeToInclusive<i32> { return ...1; }
[01:22:25]    |                                                        ^^^
[01:22:25]    |
[01:22:25]    = help: Use `..` if you need an exclusive range (a < b)
[01:22:25]    = help: or `..=` if you need an inclusive range (a <= b)
[01:22:25] 
[01:22:25] error: internal compiler error: unexpected panic
[01:22:25] 
[01:22:25] note: the compiler unexpectedly panicked. this is a bug.
[01:22:25] 
[01:22:25] note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
[01:22:25] 
[01:22:25] note: rustc 1.23.0-dev running on x86_64-unknown-linux-gnu
[01:22:25] 
[01:22:25] thread 'rustc' panicked at 'parse_prefix_range_expr: token DotDotDot is not DotDot/DotDotEq', /checkout/src/libsyntax/parse/parser.rs:3020:8
[01:22:25] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:22:25] 
[01:22:25] 
[01:22:25] ------------------------------------------
[01:22:25] 
[01:22:25] thread '[pretty] run-pass/range_inclusive.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2497:8
[01:22:25] note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

</details>

<details><summary>run-pass/range_inclusive_gate.rs</summary>

```
[01:22:25] ---- [pretty] run-pass/range_inclusive_gate.rs stdout ----
[01:22:25] 	
[01:22:25] error: pretty-printing failed in round 1 revision None
[01:22:25] status: exit code: 101
[01:22:25] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "-" "-Zunstable-options" "--unpretty" "normal" "--target" "x86_64-unknown-linux-gnu" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/range_inclusive_gate.stage2-x86_64-unknown-linux-gnu.pretty.aux" "-Crpath" "-O" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
[01:22:25] stdout:
[01:22:25] ------------------------------------------
[01:22:25] // Copyright 2016 The Rust Project Developers. See the COPYRIGHT
[01:22:25] // file at the top-level directory of this distribution and at
[01:22:25] // http://rust-lang.org/COPYRIGHT.
[01:22:25] //
[01:22:25] // Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
[01:22:25] // http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
[01:22:25] // <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
[01:22:25] // option. This file may not be copied, modified, or distributed
[01:22:25] // except according to those terms.
[01:22:25] 
[01:22:25] // Test that you only need the syntax gate if you don't mention the structs.
[01:22:25] 
[01:22:25] #![feature(inclusive_range_syntax)]
[01:22:25] 
[01:22:25] fn main() {
[01:22:25]     let mut count = 0;
[01:22:25]     for i in 0_usize...10 { assert!(i >= 0 && i <= 10); count += i; }
[01:22:25]     assert_eq!(count , 55);
[01:22:25] }
[01:22:25] 
[01:22:25] 
[01:22:25] ------------------------------------------
[01:22:25] stderr:
[01:22:25] ------------------------------------------
[01:22:25] error: `...` syntax cannot be used in expressions
[01:22:25]   --> <anon>:17:21
[01:22:25]    |
[01:22:25] 17 |     for i in 0_usize...10 { assert!(i >= 0 && i <= 10); count += i; }
[01:22:25]    |                     ^^^
[01:22:25]    |
[01:22:25]    = help: Use `..` if you need an exclusive range (a < b)
[01:22:25]    = help: or `..=` if you need an inclusive range (a <= b)
[01:22:25] 
[01:22:25] error: aborting due to previous error
```

</details>

**Comment by Badel2** [comment]
I'm confused. The error says that line 20 of range_inclusive.rs is `...` when it was changed to `..=`:
https://github.com/rust-lang/rust/blob/54bbd567153463ce81d3ff7f0ca980a7f73cd3af/src/test/run-pass/range_inclusive.rs#L20

**Comment by petrochenkov** [comment]
@Badel2 
The error may refer to the pretty-printed version of `range_inclusive.rs` which still may have `...` if some code in pretty-printing wasn't updated.

Could you also add a test checking that if `...` is actually written (`...b` and `a ... b`), then recovery works and no ICE happens.

**Comment by Badel2** [comment]
I have added a parse-fail test, but I can't find the pretty-printing code. The closest I got is the code for ranges in patterns, but I can't find the one for expressions. https://github.com/rust-lang/rust/blob/19402f11e12042367f2b711155cdde61a608b660/src/librustc/hir/print.rs#L1821

**Comment by petrochenkov** [comment]
@Badel2 
That's HIR pretty-printing.
There's also AST pretty-printing in `libsyntax/print/pprust.rs` and it's AST pretty-printing which is tested by `pretty` tests.
This line in particular looks suspicious https://github.com/rust-lang/rust/blob/19402f11e12042367f2b711155cdde61a608b660/src/libsyntax/print/pprust.rs#L2207

**Comment by Badel2** [comment]
@petrochenkov Thanks! I hope it's fixed now.

**Comment by Badel2** [comment]
That's strange, in my machine I run `./x.py test src/test/parse-fail/` and it works... According to the logs it only detected the first error. I have no idea why :)

**Comment by petrochenkov** [comment]
I see why only the first error is detected, the test is missing
```
// compile-flags: -Z parse-only -Z continue-parse-after-error
```
, but it should fail locally as well.

(`-Z continue-parse-after-error` allows reporting multiple errors during parsing phase, `-Z parse-only` is the only thing that differs `parse-fail` tests from `compile-fail` tests (less work for the compiler)).

**Comment by Badel2** [comment]
@petrochenkov Same result, it fails on travis but it works on my machine. I haven't run `./x.py clean` since the first commit 3 days ago, so maybe I'm using an outdated tool or something.

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -3008,22 +3008,22 @@ impl<'a> Parser<'a> {
...
                                -> PResult<'a, P<Expr>> {
-        // SNAP remove DotDotDot
-        debug_assert!([token::DotDot, token::DotDotDot, token::DotDotEq].contains(&self.token),
-                      "parse_prefix_range_expr: token {:?} is not DotDot/DotDotDot/DotDotEq",
+        // Check for deprecated `...` syntax
+        if self.token == token::DotDotDot {
+            self.err_dotdotdot_syntax(self.span);
+        }
+
+        debug_assert!([token::DotDot, token::DotDotEq].contains(&self.token),
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: It looks like this `debug_assert` may be firing on CI.
IIRC, these kinds of asserts are enabled on PR merge, but disabled by default for local builds.

Comment (Badel2) [reply src/libsyntax/parse/parser.rs]: Ok, for some reason I thought the code would stop executing after the err.emit(). Now it all makes sense. Thank you very much!

**Pull request merged**